### PR TITLE
Accelerate sample to counts conversion.

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
While doing work on mid-circuit measurements #5088 #5120 , I found `_samples_to_counts` to be a bottleneck in small circuits with a lot of shots (>1e5). For example, the circuit below spends:

- 1.9 sec. in `_samples_to_counts` on `master`
- 0.09 sec. in `_samples_to_counts` on `optim/_samples_to_counts`
 
```
import pennylane as qml
import pennylane.numpy as np
dev = qml.device("default.qubit", shots=100000)
@qml.qnode(dev)
@qml.defer_measurements
def func2(x, y):
    qml.RX(x, wires=0)
    m0 = qml.measure(0)
    qml.RX(y, wires=1)
    m1 = qml.measure(1)
    return qml.counts(wires=[0, 1])
results2 = func2(np.pi / 4, np.pi / 4)
```

**Description of the Change:**
Accelerate NaN pruning and int to str conversion in `_samples_to_counts`. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/measurements/counts.py [hunk 1]
```diff
@@ -301,17 +302,30 @@ def circuit(x):
 
         if self.obs is None and not isinstance(self.mv, MeasurementValue):
             # convert samples and outcomes (if using) from arrays to str for dict keys
-            samples = qml.math.array(
-                [sample for sample in samples if not qml.math.any(qml.math.isnan(sample))]
-            )
+
+            # remove nans
+            mask = qml.math.isnan(samples)
+            nsam = shape[-1]
```
Comment (timmysilv) [inline pennylane/measurements/counts.py:308]: can you call this variable `num_wires` instead? `nsam` isn't very informative

File: pennylane/measurements/counts.py [hunk 2]
```diff
@@ -301,17 +302,30 @@ def circuit(x):
...
+
+            # convert to string
+            def convert(x):
+                return f"{x:064b}"[-nsam:]
+
+            exp2 = 2 ** np.arange(nsam - 1, -1, -1)
+            samples = np.einsum("...i,i", samples, exp2)
+            new_shape = samples.shape
             samples = qml.math.cast_like(samples, qml.math.int8(0))
-            samples = qml.math.apply_along_axis(_sample_to_str, -1, samples)
```
Comment (timmysilv) [inline pennylane/measurements/counts.py:308]: please delete the unused `_sample_to_str` function above

**Review by timmysilv — APPROVED** [review]
amazing!! left two little comments to address, but happy to see this merged once those changes are in

File: pennylane/measurements/counts.py (hunk 2 above)
Comment (vincentmr) [reply pennylane/measurements/counts.py:308]: Hadn't realized this isn't used anymore, thanks.

File: pennylane/measurements/counts.py
[the remark this answers is not in the document]
Comment (astralcai) [reply pennylane/measurements/counts.py:1]: It's probably just `numpy` doing masking faster than Python list comprehension?

[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/measurements/counts.py:1]: I don't know whether this is an absolute statement, but yes, Numpy is usually faster than list comprehensions. That's it for the NaN finding. For the list-to-str conversion, I think we have versus:

- Casting list elements to str + concatenating strings.
- Matrix-vector product + using f-string to represent integers in binary format.

The second option could be made faster if we could somehow change `f"{x:064b}"[-num_wires:]` for `f"{x:0<num_wires>b}"`.

[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/measurements/counts.py:1]: > The second option could be made faster if we could somehow change `f"{x:064b}"[-num_wires:]` for `f"{x:0<num_wires>b}"`.

I think that's straightforward actually. Let me make that change.

**Review by albi3ro — APPROVED** [review]
🚀

**Pull request merged**